packages/block-library/src/image/edit.js

```diff
@@ -236,27 +284,129 @@ class ImageEdit extends Component {
 	}
 
 	updateAlignment( nextAlign ) {
-		const extraUpdatedAttributes = [ 'wide', 'full' ].indexOf( nextAlign ) !== -1 ?
-			{ width: undefined, height: undefined } :
-			{};
-		this.props.setAttributes( { ...extraUpdatedAttributes, align: nextAlign } );
+		if ( nextAlign === 'wide' || nextAlign === 'full' ) {
+			// Reset all sizing attributes.
+			this.resetWidthHeight();
+		}
+
+		this.props.setAttributes( { align: nextAlign } );
+	}
+
+	updateImageURL( url, imageSizeOptions ) {
+		this.resetWidthHeight();
+		let fileWidth;
+		let fileHeight;
+
+		if ( imageSizeOptions.width && imageSizeOptions.height ) {
+			fileWidth = imageSizeOptions.width;
+			fileHeight = imageSizeOptions.height;
+		} else {
+			// Find the image data.
+			map( imageSizeOptions, ( { value, imageData } ) => {
+				if ( value === url ) {
+					fileWidth = imageData.width;
+					fileHeight = imageData.height;
+					return false;
+				}
+			} );
+		}
+
+		this.props.setAttributes( {
+			url,
+			fileWidth,
+			fileHeight,
+		} );
+	}
+
+	updateWidth( width, fileWidth, fileHeight, userSet ) {
+		width = parseInt( width, 10 );
+
+		// Reset the image size when the user deletes the value.
+		if ( ! width || ! fileWidth || ! fileHeight ) {
+			this.resetWidthHeight();
+			return;
+		}
+
+		const height = round( fileHeight * ( width / fileWidth ) );
+		this.setWidthHeight( width, height, fileWidth, fileHeight, userSet );
+	}
+
+	updateHeight( height, fileWidth, fileHeight, userSet ) {
+		height = parseInt( height, 10 );
+
+		// Reset the image size when the user deletes the value.
+		if ( ! height || ! fileWidth || ! fileHeight ) {
+			this.resetWidthHeight();
+			return;
+		}
+
+		const width = round( fileWidth * ( height / fileHeight ) );
+		this.setWidthHeight( width, height, fileWidth, fileHeight, userSet );
 	}
 
-	updateImageURL( url ) {
-		this.props.setAttributes( { url, width: undefined, height: undefined } );
+	setWidthHeight( width, height, fileWidth, fileHeight, userSet ) {
+		// Set image size and also whether set directly by the user.
+		userSet = userSet || undefined;
```

Comment:
Is this needed?

Reply:
if this is what I'm thinking it is, it sounds like it's something like `intentionallySetWidth` and `intentionallySetHeight` or `displayWidth` and `displayHeight`

is there confusion since this boolean value doesn't entirely describe why it exists or what the user set?

Reply:
You mean `userSet = userSet || undefined;` or `userSet`? The first is probably not, just some "normalization", but as that's an internal function it can be skipped.

The `userSet` indicates whether the image `width` and `height` were set directly by the user (typed in the width or height fields in the inspector), or they were added as a result of resizing the image by dragging. This would make a difference when the theme or a plugin is deciding to scale or not to scale an image on the front-end.

In the previous iteration these were actually two values: `userWidth` and `userHeight`. Switched it to one boolean after some critique that we store too much image meta in the block props. Can revert that easily :)

Reply:
Worth noting there's a subtlety here in that if the argument was called with `userSet = false`, it would normalize the value to `undefined`, which in practice has the same result since the default of the attribute is `false`. And while that default is unlikely to change, **if** it were to, the normalizing would have the adverse effect of preventing a value of explicit `false` to be provided as an argument.